js/node_shared/unix.js

```diff
@@ -0,0 +1,369 @@
+const os = require("os");
+const fs = require("fs");
+const process = require("process");
+
+globalThis.spawnSync = spawnSync;
+globalThis.fs = fs;
+globalThis.os = os;
+globalThis.process = process;
+
+// mapping pid to its own process
+globalThis.pidToProcessTable = new Map();
+// mapping pid of parent to its children
+globalThis.pidToChildrenTable = new Map();
+// There will be collisions if we ever allocate 20000 or more file descriptors
+// "naturally", that is, through `caml_sys_open`.
+globalThis.fdCount = 20000;
+globalThis.fdTable = new Map();
+
+//Provides: unix_spawn
+function unix_spawn(executable, args, optenv, usepath, redirect) {
+  console.error("unix_spawn");
+  console.error(executable);
+  console.error(args);
+  console.error(optenv);
+  console.error(usepath);
+  console.error(redirect);
+
+  // get rid of the mandatory tag and the first argument, which is the
+  // executable path
+  let argv = args.slice(2);
+
+  // These redirects are given to `unix_spawn` by Bos via `Unix.create_process_gen`,
+  // which gets them from Bos explicitly calling `pipe`.
+  // Via our convention in unix_pipe, these are not actually "real" file descriptors,
+  // but fresh identifiers we will use to map to the output of the corresponding
+  // child process.
+  // As such, let's just use the `pipe` setting for the stdio of this child process.
+  let output_fd = redirect[2];
+
+  /* In Javascript, there is no `execvp` and `execv` which
+     switch between searching in the PATH or not.
+     We will just run the vanilla `exec`, which should behave
+     as an `execvp`.
+     This might have slightly differing behavior than the
+     actual implementation.
+   */
+  if (optenv == 0) {
+    const child = globalThis.spawnSync(executable, argv, { stdio: "pipe" });
+
+    globalThis.pidToProcessTable.set(child.pid, child);
+    globalThis.fdTable.set(output_fd, child.stdout);
+
+    return child.pid;
+  } else {
+    let env = {};
+    for (const arg of optenv[1].slice(1)) {
+      const [key, value] = arg.split("=");
+      env[key] = value;
+    }
+    const child = globalThis.spawnSync(executable, argv, {
+      env: env,
+      stdio: "pipe",
+    });
+
+    globalThis.pidToProcessTable.set(child.pid, child);
+    globalThis.fdTable.set(output_fd, child.stdout);
+
+    let current_pid = globalThis.process.pid;
+
+    // associate parent to new child
+    if (globalThis.pidToChildrenTable.has(current_pid)) {
+      globalThis.pidToChildrenTable.set(
+        current_pid,
+        globalThis.pidToChildrenTable.get(current_pid).concat([child])
+      );
+    } else {
+      globalThis.pidToChildrenTable.set(current_pid, [child]);
+    }
+
+    return child.pid;
+  }
+}
+
+//Provides: unix_waitpid
+function unix_waitpid(flags, pid_req) {
+  console.error("unix_waitpid");
+  console.error(flags);
+  console.error(pid_req);
+
+  // TODO: flags stuff is not implemented
+  // TODO: WIFSTOPPED is not currently possible, because I don't know how to
```

Comment:
there's child.on('exit') and similarly on close see [here](https://stackoverflow.com/questions/34881343/node-js-detect-a-child-process-exit) using spawnSync ensures the process will have stopped before this code ever runs though

Reply:
is this OK to leave as a TODO, then, since I am using the synchronous spawn?

Reply:
Yea, it only matters if we use async spawn